### PR TITLE
Extract metadata from share, use spec share

### PR DESCRIPTION
**Pull request by nkryuchkov**
The PR introduces a new type `SSVShare` instead of `beacon.Share` which contains `spectypes.Share` and some share metadata from `beacon.Share`

### CODE REVIEW

File: operator_v2/validator/controller.go [hunk 1]
```diff
@@ -310,29 +327,29 @@ func (c *controller) ListenToEth1Events(feed *event.Feed) {
...
-	c.logger.Info("starting validators setup...", zap.Int("shares count", len(shares)))
+func (c *controller) setupValidators(shareList []*share.Share, metadataList []*share.Metadata) {
+	c.logger.Info("starting validators setup...", zap.Int("shares count", len(shareList)))
 	var started int
 	var errs []error
 	var fetchMetadata [][]byte
-	for _, validatorShare := range shares {
-		v := c.validatorsMap.GetOrCreateValidator(validatorShare)
+	for i, validatorShare := range shareList {
+		v := c.validatorsMap.GetOrCreateValidator(validatorShare, metadataList[i])
```
Comment (nivBlox) [inline operator_v2/validator/controller.go:349]: Add test

File: operator_v2/validator/share_opt.go [hunk 2]
```diff
@@ -21,56 +22,109 @@ type ShareOptions struct {
...
-		return &share, nil
 	}
-	return nil, errors.New("empty or invalid share")
+
+	var operators [][]byte
+	for _, op := range options.Operators {
+		operators = append(operators, []byte(op))
+	}
+
+	var operatorIDs []uint64
```
Comment (nivBlox) [inline operator_v2/validator/share_opt.go:114]: we still need operators id? could maybe remove in sec pr

File: operator_v2/validator/controller.go [hunk 3]
```diff
@@ -207,26 +207,26 @@ func (c *controller) setupNetworkHandlers() error {
...
 }
 
-func (c *controller) GetAllValidatorShares() ([]*beaconprotocol.Share, error) {
+func (c *controller) GetAllValidatorShares() ([]*share.Share, error) {
 	return c.collection.GetAllValidatorShares()
 }
 
 func (c *controller) GetValidatorStats() (uint64, uint64, uint64, error) {
-	allShares, err := c.collection.GetAllValidatorShares()
+	allMetadata, err := c.collection.GetAllValidatorMetadata()
```
Comment (nivBlox) [inline operator_v2/validator/controller.go:215]: Let's use 1 function for saving and fetch. also lets wrap both share and metadata under one struct

File: protocol/v2/types/ssvshare.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+// SetOperators sets operators public keys.
+func (s *SSVShare) SetOperators(pks [][]byte) {
+	s.Operators = make([][]byte, len(pks))
+	copy(s.Operators, pks)
+}
+
+// ShareMetadata represents Metadata of SSVShare.
+type ShareMetadata struct {
```
Comment (lior-blox) [inline protocol/v2/types/ssvshare.go:72]: @nkryuchkov lets call it Metadata

File: protocol/v2/types/ssvshare.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	Stats        *beaconprotocol.ValidatorMetadata
```
Comment (lior-blox) [inline protocol/v2/types/ssvshare.go:73]: change stats to beaconMetadata

File: operator_v2/validator/event_handler.go [hunk 6]
```diff
@@ -111,17 +110,19 @@ func (c *controller) handleOperatorRemovalEvent(
 		return logFields, nil
 	}
 
-	shares, err := c.collection.GetOperatorIDValidatorShares(event.OperatorId, false)
+	shares, err := c.collection.GetFilteredValidatorShares(func(share *types.SSVShare) bool {
```
Comment (nivBlox) [inline operator_v2/validator/event_handler.go:113]: Add constant filters functions for reuse

File: protocol/v2/ssv/validator/validator.go [hunk 7]
```diff
@@ -2,30 +2,29 @@ package validator
...
 )
 
 type Options struct {
 	Logger      *zap.Logger
 	Network     qbft.Network
 	Beacon      ssv.BeaconNode
 	Storage     qbft.Storage
-	Share       *beacon.Share
-	Signer      types.KeyManager
+	Share       *types.SSVShare
```
Comment (lior-blox) [inline protocol/v2/ssv/validator/validator.go:26]: rename to SSVShare

File: operator_v2/validator/controller_shares.go [hunk 8]
```diff
@@ -45,24 +45,28 @@ func (c *controller) loadShare(options ShareOptions) (string, error) {
...
+
+	if err := c.keyManager.AddShare(shareKey); err != nil {
+		return "", errors.Wrap(err, "could not save share key from share options")
+	}
+	if err := c.collection.SaveValidatorShare(share); err != nil {
+		return "", errors.Wrap(err, "could not save share from share options")
 	}
 
-	return "", errors.New("returned nil share")
+	return options.PublicKey, err
```
Comment (lior-blox) [inline operator_v2/validator/controller_shares.go:71]: return the share

**Comment by amir-blox** [comment]
- Add more tests for share
- Add tests for validator controller: what happens we load a validator with nil share (nil metadata)

File: operator_v2/validator/share_opt.go [hunk 9]
```diff
@@ -21,56 +22,75 @@ type ShareOptions struct {
 	OperatorIds  []int          `yaml:"OperatorIds" env:"LOCAL_OPERATOR_IDS" env-description:"Local validator selected operator ids"`
```
Comment (lior-blox) [inline operator_v2/validator/share_opt.go:22]: Operators & OperatorIds should be removed cause we can extract the data from the commitee

Comment (lior-blox) [reply operator_v2/validator/share_opt.go:22]: remove not used ShareKey as well

File: operator_v2/validator/controller_shares.go [hunk 10]
```diff
@@ -31,38 +35,42 @@ func (c *controller) loadSharesFromConfig(items []ShareOptions) {
 				c.logger.Error("failed to load validator share data from config", zap.Error(err))
 				continue
 			}
-			addedValidators = append(addedValidators, pubkey)
+			addedValidators = append(addedValidators, hex.EncodeToString(pubkey.ValidatorPubKey))
```
Comment (nkryuchkov) [inline operator_v2/validator/controller_shares.go:38]: rename `pubkey` to `share`

File: operator_v2/validator/controller_shares.go [hunk 11]
```diff
@@ -31,38 +35,42 @@ func (c *controller) loadSharesFromConfig(items []ShareOptions) {
...
 	share, err := options.ToShare()
 	if err != nil {
-		return "", errors.WithMessage(err, "failed to create share object")
+		return nil, errors.WithMessage(err, "failed to create share object")
 	}
+
 	shareKey := &bls.SecretKey{}
-	if err = shareKey.SetHexString(options.ShareKey); err != nil {
-		return "", errors.Wrap(err, "failed to set hex private key")
+	if err = shareKey.SetLittleEndian(share.SharePubKey); err != nil {
```
Comment (nkryuchkov) [inline operator_v2/validator/controller_shares.go:54]: revert

File: operator_v2/validator/controller_shares.go [hunk 12]
```diff
@@ -31,38 +35,42 @@ func (c *controller) loadSharesFromConfig(items []ShareOptions) {
...
+		return nil, errors.New("returned nil share")
+	}
+
+	if updated, err := UpdateShareMetadata(share, c.beacon); err != nil {
+		return nil, errors.Wrap(err, "could not update share stats")
+	} else if !updated {
+		return nil, errors.New("could not find validator metadata")
+	}
+
+	if err := c.keyManager.AddShare(shareKey); err != nil {
```
Comment (nkryuchkov) [inline operator_v2/validator/controller_shares.go:68]: make sure that correct share key is added (private key from yaml config)

File: operator_v2/validator/storage_test.go [hunk 13]
```diff
@@ -26,19 +29,23 @@ func TestValidatorSerializer(t *testing.T) {
 	require.NoError(t, err)
 
 	validatorShare, _ := generateRandomValidatorShare(splitKeys)
-	b, err := validatorShare.Serialize()
+	b, err := validatorShare.Encode()
 	require.NoError(t, err)
 
-	obj := basedb.Obj{
-		Key:   validatorShare.PublicKey.Serialize(),
+	obj1 := basedb.Obj{
```
Comment (nkryuchkov) [inline operator_v2/validator/storage_test.go:35]: revert to obj

File: protocol/v2/types/ssvshare.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+	spectypes.Share
+	Metadata
+}
+
+// Encode encodes SSVShare using gob.
+func (s *SSVShare) Encode() ([]byte, error) {
+	var b bytes.Buffer
+	e := gob.NewEncoder(&b)
+	if err := e.Encode(s); err != nil {
+		return nil, fmt.Errorf("failed to encode serializedValidator: %w", err)
```
Comment (nkryuchkov) [inline protocol/v2/types/ssvshare.go:24]: fix error text

File: protocol/v2/types/ssvshare.go [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	return b.Bytes(), nil
+}
+
+// Decode decodes SSVShare using gob.
+func (s *SSVShare) Decode(data []byte) error {
+	d := gob.NewDecoder(bytes.NewReader(data))
+	if err := d.Decode(s); err != nil {
+		return fmt.Errorf("failed to get val value: %w", err)
```
Comment (nkryuchkov) [inline protocol/v2/types/ssvshare.go:34]: fix error text

**Review by amir-blox — APPROVED** [review]

**Review by nivBlox — APPROVED** [review]
